dnsprovider/pkg/dnsprovider/providers/scaleway/dns.go

```diff
@@ -0,0 +1,571 @@
+package dns
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"golang.org/x/oauth2"
+	"io"
+	"k8s.io/klog/v2"
+	kopsv "k8s.io/kops"
+	"k8s.io/kops/dns-controller/pkg/dns"
+	"k8s.io/kops/dnsprovider/pkg/dnsprovider"
+	"k8s.io/kops/dnsprovider/pkg/dnsprovider/rrstype"
+	"os"
+
+	"github.com/scaleway/scaleway-sdk-go/api/domain/v2beta1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+var _ dnsprovider.Interface = Interface{}
+
+const (
+	ProviderName   = "scaleway"
+	KopsDomainName = "scaleway-terraform.com" // TODO: replace with real domain name
+)
+
+func init() {
+	dnsprovider.RegisterDNSProvider(ProviderName, func(config io.Reader) (dnsprovider.Interface, error) {
+		client, err := newClient()
+		if err != nil {
+			return nil, err
+		}
+
+		return NewProvider(client, KopsDomainName), nil
+	})
+}
+
+// TokenSource implements oauth2.TokenSource
+type TokenSource struct {
+	AccessToken string
+}
+
+// Token returns oauth2.Token
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+func newClient() (*scw.Client, error) {
+	accessToken := os.Getenv("SCW_ACCESS_KEY")
+	if accessToken == "" {
+		return nil, errors.New("SCW_ACCESS_KEY is required")
+	}
+	tokenSource := &TokenSource{
+		AccessToken: accessToken,
+	}
+	oauthClient := oauth2.NewClient(context.TODO(), tokenSource)
+
+	// TODO: check if it's necessary to have both the oauth token & the scw env
```

Comment:
Usually when you have different ways of getting the information you need to merge the different input ways (command line, an environment variable, ...) we need to define a priority order.